### ISSUE TITLE
[BUG] `==` and `=` in declarations are confusing.

### ISSUE THREAD

**Issue by msadeqhe** [hsutter/cppfront#824]
# Description

Considering @hsutter's comment from discussion #623:

> Thanks! I think I've answered this, and addressed it for `<<`'s interactions with comparisons, in the comment thread starting here: [#817 (comment)](https://github.com/hsutter/cppfront/issues/817#issuecomment-1807398464)
> 
> Briefly: It's reasonable for `<<` with its built-in bitwise meaning to be high-precedence, but in the 1990s overuse of overloading was popular, including its use for streaming I/O which inherently wants to be low-precedence but can't change the actual operator's precedence... and that overloaded use of `<<` became popular so we keep having this surprise. In fact, I think `<<` is the poster child for precedence issues, because it's the major case that invented a later use of the operator for something very different, that became very popular.
> 
> Operator overloading is fine, but I view `<<` for streaming as a warning example about why we shouldn't overuse the feature, and when we do use it we should closely follow the built-in operators' meanings... [as Scott Meyers would say (and did), "do as the ints do."](https://www.google.com/search?q=scott+meyers+%22do+as+the+ints+do%22)

The part _"we should closely follow the built-in operators' meanings"_ is important here.

# To Reproduce

## Notation `==`

Considering type and namespace aliases:

```cpp
v32: type == std::vector<i32>;
```

A typical C++ programmer expects `==` to return the result of a comparison, but it sets `v32` type definition from another type instead. Therefore `==` is an assignment operator in declarations, but it is a comparison operator in expressions.

Considering `constexpr` functions and variables, `==` is more confusing:

```cpp
fnc1: () -> bool == (x == y); // =='s are not the same!
fnc2: () -> bool == x == y; // Should it be allowed?!
```

The first `==` sets `fnc1` function definition, but the second `==` compares two literals and returns the result. A typical programmer may ask why it's happening while they are right to each other.

Considering `concept` declarations in Cpp2:

```cpp
arithmetic: <T> concept = std::integral<T> || std::floating_point<T>;
```

Unlike types and namespaces, it uses `=` instead of `==`. Cpp2 is/was designed for C++ programmers of which they are more familiar with declaring stuff by `=` than `==` in Cpp1:

```cpp
using v32 = std::vector<i32>;
template<class T> concept yes = true;
```

## Notation `=`

Considering function declaration syntax:

```cpp
fnc1: () = {}
fnc2: () = something();
```

If we ask programmers who are not familiar with Cpp2 that _"What does `fnc1` function do?"_, they probably will answer `fnc1` returns an empty list `{}`. Because that's natural to programmers. `=` puts the result of its right side to its left side. `{}` is on the righ side, so it must has a value.

But in Cpp2 this known behavior doesn't work. `=` doesn't set anything in the example, and `{}` doesn't have a value too! Of course `void` is nothing, it's not even similar to `null` and `nullptr` which they are values.

Looking at `fnc2`, a typical programmer which is not familiar with Cpp2, thinks `something()` has a value and the function returns the value. Although we had this natural behavior in Cpp2 before, but @hsutter changed this behavior (because of issue #257) because of the push-back he got on issues/comments as he explained it [here](https://github.com/hsutter/cppfront/discussions/714#discussioncomment-7158899).

Why not bring back the original behavior for function declarations with `= expression`?

# Additional context

Cpp2 may have a simple defined rule to separate statements from expressions in declarations, if one extra `=` could be dropped from them:

```cpp
fnc1: () { print(10); } // A block statement
fnc2: () = 10; // An expression
```

`fnc1` is defined by a block statement, because it doesn't have `=`, but `fnc2` is defined by an expression, because it has `=`. This rule would allow to categorize function decleration to two syntax:

First, the function declaration (also lambdas) with block statement, has not a return type by default:

```cpp
 fnc1: () -> void { print(0); }
 fnc1: () { print(0); }

call(: () { print(0); }); // Lambda
```

Second, the function declaration (also lambdas) with an expression after `=`, has a generic return type by default:

```cpp
 fnc1: () -> _ = 10;
 fnc1: () = 10;

call(: () = 10); // Lambda
```

In a similar manner, types and namespaces would be (more familiar to C++ programmers):

```cpp
cls1: type { ... }
cls2: type = cls1;

nsp1: namespace { ... }
nsp2: namespace = nsp1;
```

In general `{}` is a block statement and `= ...` is an expression or a name (or a fully qualified name).

Also for `constexpr` functions and variables, other solutions such as using a keyword can be considered.

Discussions #742 and #714 are related to this topic.

**Comment by msadeqhe** [comment]
The unifying functions and blocks as explained by @hsutter from [here](https://github.com/hsutter/cppfront/issues/386#issuecomment-1518194517), would be changed like this:

```cpp
f: (x: int = init) { ... }    // x is a parameter to the function
f: (x: int = init) statement; // same, { } is implicit
f: (x: int = init) = expr;    // { return expr; } is implicit

 : (x: int = init) { ... }    // x is a parameter to the lambda
 : (x: int = init) statement; // same, { } is implicit
 : (x: int = init) = expr;    // { return expr; } is implicit

   (x: int = init) { ... }    // x is a “let parameter” to the block
   (x: int = init) statement; // same, { } is implicit

                   { ... }    // x is a “let parameter” to the block
                   statement; // same, { } is implicit
```

As we see, terser lambda syntax `: (x) x + 1` breaks the rules, because `x + 1` is not a statement, it's an expression. By making `=` to be required in lambda syntax, `: (x) = x + 1` may seems to be a better choice, because `=` improves its readability and we can specify the return type if necessary.

**Comment by JohelEGP** [comment]
Other related discussions: #634, #761.

> As we see, terser lambda syntax `: (x) x + 1` breaks the rules, because `x + 1` is not a statement, it's an expression.

Up to date: <https://github.com/hsutter/cppfront/wiki/Design-note%3A-Defaults-are-one-way-to-say-the-same-thing#from-named-functions-to-lambdas-to-parameterized-blocksstatements-to-ordinary-blocksstatements>:
> ```
> f:(x: int = init) = { ... }     // x is a parameter to the function
> f:(x: int = init) = statement;  // same, { } is implicit
>
>  :(x: int = init) = { ... }     // x is a parameter to the lambda
>  :(x: int = init) = statement;  // same, { } is implicit
>
>   (x: int = init)   { ... }     // x is a parameter to the block
>   (x: int = init)   statement;  // same, { } is implicit
>
>                     { ... }     // an ordinary block
>                     statement;  // same, { } is implicit
> ```

I think `:(x) x + 1` is actually a fine,
but separate case from explaining blocks.
That's why the design note was created for, after all (#714).
(In another section further down:)
> To me, allowing a generic function `f:(i:_) -> _ = { return i+1; }` to be spelled `f:(i) i+1;` is like that...

**Comment by realgdman** [comment]
It is statement (in layman word, it states a fact) that left side equals to right side.
I don't think it's a problem, `=` has meaning like a `a <- b` i.e. a was something, then it changes value to b. That has sense in runtime context, when variable, well, can vary mutable. It doesn't have much sense at compile time context, because we usually cannot modify or remove things from program, only add facts about it, but after compile, all those facts are set in stone and cannot be changed, so `=` is moot. That's all humble IMO of course.

**Comment by JohelEGP** [comment]
> It doesn't have much sense at compile time context, because we usually cannot modify or remove things from program, only add facts about it, but after compile, all those facts are set in stone and cannot be changed, so `=` is moot.

The same is true of `const` and `constexpr` variables.
Would you say those use `=` for consistency?

**Comment by msadeqhe** [comment]
@realgdman The `a <- b` relation doesn't work well with `= {}`. For example:

```cpp
fnc1: () = anything = {}  // ERROR
fnc2: () = {} = anything; // ERRORR
```

`{}` cannot be set to anything, also we cannot set it to anything, because `{}` is not an expression, it`s a block statement.

On the other hand:

```cpp
fnc1: () = x = y; // OK
```

That's because `=` (as most programmers except it that way) works with expressions.

**Comment by msadeqhe** [comment]
@realgdman The `a <- b` relation doesn't work well with `= {}`. For example:

```cpp
fnc1: () = anything = {}  // ERROR
fnc2: () = {} = anything; // ERROR

fnc3: () -> i32 = x = { return y; }  // ERROR
fnc4: () -> i32 = { return x; } = y; // ERROR
```

And in general, `= {} ` doesn't work consistently in other parts of the language:

```cpp
x: = something;
x = { /* ... */ }; // ERROR
x = another;       // OK
```

`{}` cannot be set to anything, also we cannot set it to anything, because `{}` is not an expression, it`s a block statement.

On the other hand:

```cpp
fnc1: () = x = y; // OK
```

That's because `=` (as most programmers except it that way) works with expressions.

**Comment by realgdman** [comment]
Does that works because `=` assignment returns/has a value? It is natural for C/C++, but there are languages where assignment doesn't return value. Also returning value from `=` doesn't have much sense at global compile time level? Or do you mean need of  chaining v32: type == myT : type == std::vector<i32>

**Comment by msadeqhe** [comment]
No, I mean:

```cpp
x: TYPE = 10;
x = { /* ... */ }; // Syntax error
```

Generally `{}` doesn't work with `=` in Cpp2. Because it's meaningless to put a block statement in a side of assignment.

**Comment by msadeqhe** [comment]
Let's explore other concepts of `{}` in programming.

### Block Expression

`{}` can be a block expression, but we don't have it yet in Cpp2. Considering *`if`-expression*:

```cpp
var1: = if x < 10 { x } else { x - 10 };
```

Those `{}`'s are not block statements, they are block expressions. So it's meaningful to put them within expression in a side of assignment.

## List of Items

Also `{}` can be a list of items:

```cpp
var1: std::vector<i32> = {1, 2, 3};
```

That `{}` is not a block statement, it's an expression. Its value is a list of items.

## Conclusion (Lambda Syntax Ambiguity)

Considering those two concepts and function declaration syntax, if Cpp2 will support them, we will have:

```cpp
// `{}` is a block statement.
fnc1: () -> std::vector<i32> = { /*...*/ }

// `{}` is a block expression or a list of items.
fnc2: () -> std::vector<i32> = { /*...*/ };
```

The meaning of `{}` is confusing. It's meaning can be changed with a single `;` at the end. Now, let's consider the lambda syntax:

```cpp
// `{}` is a block statement.
call(: () -> std::vector<i32> = { /*...*/ });

// `{}` is a block expression or a list of items.
call(: () -> std::vector<i32> = { /*...*/ });
```

Lambda syntax doesn't have `;` at the end. So both of them looks the same now. Cpp2 has to look inside `{}` to find what it means. It's ambiguous, and it harms both readability and toolability.

If @hsutter would approve this bug report with the new syntax, `fnc1` declaration would be invalid. On the other hand we would have `fnc3` declaration:

```cpp
// `{}` is a block expression or a list of items.
fnc2: () -> std::vector<i32> = { /*...*/ };

// `{}` is a block statement.
fnc3: () -> std::vector<i32> { /*...*/ }
```

Obviousely `{}` is not a block expression or a list of items in `fnc2`, because it's on the right side of assignment operator, and `{}` is a block statement in `fnc3`. The lambda syntax of them is like this:

```cpp
// `{}` is a block expression or a list of items.
call(: () -> std::vector<i32> = { /*...*/ });

// `{}` is a block statement.
call(: () -> std::vector<i32> { /*...*/ });
```

Fortunately they are different, the tool can easily know what `{}` means in both of them as its readability is improved.

## Consistency with Control Structurs

The new syntax improves consistency with control structures.

For example, this is how they look in old syntax:

```cpp
if x < 10 {
    return x;
}

f: () -> i32 = {
    return 0;
}
```

Now, comparing them in new syntax:

```cpp
if x < 10 {
    return x;
}

f: () -> i32 {
    return 0;
}
```

Both of `{}` don't have `=` because they are block statements.

## Consistency with Variable Declarations

For example, this is how they look in old syntax:

```cpp
var1: i32 = n + 1;
fnc1: (x) x + 1;
```

Now, comparing them in new syntax:

```cpp
var1: i32 = n + 1;
fnc1: (x) = x + 1;
```

Both of them have `=` because they are defined with an expression. This is also true within `inspect` expressions:

```cpp
var1: = inspect x < 10 -> bool { is true = x; is _ = x - 10 };
```

`= ...;` is used, it's consistent with how we use it to declare variables and functions from expressions. Also `{}` in `inspect` is a block expression, thus it's within expression in a side of assignment.

## TL;DR

The point is, the consistency between declarations and control structurs will be improved with the new syntax, and it avoids ambiguity if Cpp2 will support block expressions and list of items with `{}`. Thanks.

**Comment by msadeqhe** [comment]
> Today, `name: signature = statement` is used for all declarations.

Today we have the following declarations:

```cpp
name: signature = block-statement
name: signature = expression;
name: parameter-list statement
name: parameter-list expression;
```

Am I right? So we don't have only one declaration syntax today. I've explained why they are confusing.

> That makes declarations themselves inconsistent.

My suggested syntax is:

```cpp
name: signature block-statement
name: signature = expression;
name: parameter-list statement
```

Why do you find it inconsistent? It improves Cpp2 code, because it makes `block-statement` and `statement` not to be be confused with `expression`. It makes assignment in `= expression;` a known notation to be used everywheregood. So `= expression;` is the only way to declare variables.

**Comment by JohelEGP** [comment]
> > Today, `name: signature = statement` is used for all declarations.
>
> Today we have the following declarations:
>
> ```c++
> name: signature = block-statement
> name: signature = statement
> name: signature = expression;
> name: parameter-list statement
> name: parameter-list expression;
> ```
>
> Am I right? So we don't have only one declaration syntax today. I've explained why they are confusing.

The first tree are covered by the second one.
The last two are <https://github.com/hsutter/cppfront/wiki/Design-note%3A-Defaults-are-one-way-to-say-the-same-thing>.
Are you also refuting defaults when `=` is involved?

> > That makes declarations themselves inconsistent.
>
> My suggested syntax is:
>
> ```c++
> name: signature block-statement
> name: signature = expression;
> name: parameter-list statement
> ```
>
> Why do you find it inconsistent?

Not all declarations follow the same syntax `name: signature = statement`.
Also, it's not clear how that supports parameterized expressions like `:(x) x`.

**Comment by msadeqhe** [comment]
> Are you also refuting defaults when = is involved?

No, they can have defaults:

> <sup>From [this comment](https://github.com/hsutter/cppfront/issues/824#issuecomment-1810071300):</sup>
> 
> _**EDIT:**_ For example, we can have the following lambdas:
> 
> ```c++
> : (x) -> void { print(x); }
> : (x) { print(x); }
> : (x) print(x)
> 
> : (x) -> _ { return x + 1; }
> 
> : (x) -> _ = x + 1
> : (x) = x + 1
> ```
> 
> `a: (x) = x + 1;` is similar to declaring a variable with `b: = n + 1;`. It improves the consistency of declarations.

> Not all declarations follow the same syntax `name: signature = statement`.
> Also, it's not clear how that supports parameterized expressions like `:(x) x`.

Yes, it doesn't treat statements and expressions to be the same. That's why I'm reporting this bug.

**Comment by AbhinavK00** [comment]
I simply think that pursuing this issue is not worth it. Personally, while I was initially of the same opinion, cpp2 syntax has grown on me. 
And I think there's some language that uses `==` for constants and the way cpp2 uses it kinda makes sense (except for constexpr functions).

**Comment by msadeqhe** [comment]
# Why Is It a Bug?

This bug report is about `= {}` in declarations. I ask to fix it by dropping `=` from them, so it leads to a new syntax.

## New Syntax

The ways we declare functions in new syntax are:

```cpp
// With new syntax
name: signature statement
name: signature = expression;
```

That's it. `statement` may be either a _block statement_ (`{ print(x); }`) or a _single statement_ (`print(x);`).

Infact the second one is a variant of the first one. I mean `= expression;` is a syntax sugar to `{ return expression; }`, so the new syntax can be just one grammar (with different default return type) instead of two grammar.

## Old Syntax

Considering this with the current syntax `= {}`:

```cpp
// Today in current Cpp2.
name: signature = something
name: parameter-list something
```

Just like the new syntax, the second is a variant of the first declaration in current Cpp2 (today).

That's it. Today (in current Cpp2) `something` is confusing to both programmers and tools, because we cannot simply find if `something` is a `statement` or an `expression`.

## Why Has It to Be Fixed?

In the current Cpp2's grammar, `something` can be either a _block statement_, a _single statement_ or an _expression_.

So why `something` is confusing? Let's consider the grammar of current Cpp2 (today) if we don't specify the return type of functions: 

```cpp
// Today in current Cpp2.
name: parameter-list = something
name: parameter-list something
```

`parameter-list` means we didn't have specified the return type in the `signature`. So the first `something` doesn't return a value, but the second `something` returns a value! It is inconsistent and confusing:

```cpp
a: (x) = fnc1(); // It doesn't return a value!
b: (x) fnc2();      // But it returns a value!
```

Please let's review it again, `fnc1()` doesn't return a value, but `fnc2()` returns a value! What?!

***It's the main reason behind of this bug report.***

Because `fnc1()` has `=` before it, so it doesn't have a value! It's absolutely inconsistent, and unfamiliar to C++ programmers. It has explicitly the reserve behavior of assignment. It's misleading, and also inconsistent with how we currently declare variables:

```cpp
fnc1: (x) = fnc1(); // It doesn't return a value!
var1: i32 = fnc2(); // But it returns a value!
```

Also `= something` without `signature` doesn't have a meaning in itself. They highly depend on the `signature` of the function, tools have to check the signature, and programmers have to read the signature. Otherwise no one will know if `something` is a _statement_ or an _expression_. So it forces use to check the `signature`.

Although it has a higher level view in the language, because the programmer writes code regardless of if it's a _statement_ or an _epxression_, but it complicates the grammar of Cpp2 for parsing, tooling and reading the code.

That's the reason I've created this issue as a bug report instead of a feature suggestion.

## What if we drop `=` from them?

On the other hand, if we drop `=` from `= something`, the grammer of Cpp2 will have a distinct syntax for each _block statement_ and _expression_. I mean:
- `= expression;` will be an expression everywhere. It will have a value everywhere.
- `{}` will be a block statement everywhere that `=` is not present before it. `{}` will be without `=` in both declarations and control flows consistently.
- Because of the distinct syntax and the well defined separation of `statement` and `expression`, the rules will be specified simpler than when we don't have that distinguish syntax. Tools and programmers can visually find the meaning of the code with the help of this distinguish syntax. This also allows to inconsistently add more features such as block expressions to the language (discussion #836).

Thanks.

**Comment by msadeqhe** [comment]
> ```cpp
> a: (x) = fnc1(); // It doesn't set/return a value!
> b: (x) fnc2();     // But it sets/returns a value!
> ```
> 
> Please let's review it again, `fnc1()` doesn't return a value, but `fnc2()` returns a value! What?! `=` in the first declaration has exactly the opposite behavior of assignment.

So we have to write it like this one instead, otherwise it would be a compiler error!

```cpp
a: (x) = _ = fnc1(); // It doesn't set/return a value!
b: (x) fnc2();         // But it sets/returns a value!
```

How much do you find it readable/consistent?

**Comment by msadeqhe** [comment]
# Why the new syntax is necessary?

To fix this bug:

> ```cpp
> a: (x) = fnc1(); // It doesn't set/return a value!
> b: (x) fnc2();     // But it sets/returns a value!
> ```
>   
> Please let's review it again, `fnc1()` doesn't return a value, but `fnc2()` returns a value! What?! `=` in the first declaration has exactly the opposite behavior of assignment. That falsely means when the return type is not specified, the programmer have to read it like the following sentence to find out the default return type:
> 
> ***"`=` is written before `fnc1()`, so it doesn't have a value!".***
> 
> It's absolutely inconsistent, and unfamiliar to C++ programmers.
> ```

We may consider the following solutions to fix the bug of the default return type.

## Solution 1

Add `=` to all declarations:

```cpp
a: (x) = fnc1();      // It doesn't set/return a value!
b: (x) = fnc2(); // ERROR! But it sets/returns a value!
```

In this way we have to make `_` to be the default return type for all declarations. So `b` declaration is a compiler error! and we have to explicitly write it like:

```cpp
a: (x) = fnc1();       // It doesn't set/return a value!
b: (x) -> void = fnc2(); // But it sets/returns a value!
```

The problem is that for functions with block statements, the default return type will be deduced too:

> <sup>@hsutter's [comment](https://github.com/hsutter/cppfront/discussions/714#discussioncomment-7231001):</sup>
> Quick ack: Making `-> _` deduced return types the default for namespace- and type-scope functions (as it is today for `auto`-declared trailing return syntax functions) feels problematic because it requires callers to be aware of the implementation details. And that's fine and useful, but it's more work at compile time and occasionally more brittle to changes in the function body, so it should be opt-in I think. The reason I made it the default for local anonymous functions (lambdas) at first was because they're inherently local and inline anyway so don't really hit the composable-at-scale issues.

This solution doesn't allow Cpp2 to have different default return types in terser function syntax when it's needed in different context. So we always have to write `=` regardless of if the function is declared with a block statement or an expression.

## Solution 2

Swap `=` in declarations:

```cpp
a: (x) fnc1(); // It doesn't set/return a value!
b: (x) = fnc2(); // But it sets/returns a value!
```

It looks natural. The default return type is natural for both of them. The programmer simply looks at `=` and says:

***Hey, it has `=` before `func2()`, so it must have a value.***

That's natural in other part's of the language. It improves readability. It makes `=` an important part of function declarations. It's consistent with how we declare variables, or assign values to them:

```cpp
b: (x) = fnc2();

x: i32 = 2;
x = 10;
```

This change leads to drop `=` from full function declaration syntax too:

```cpp
a: (x) -> void { fnc1(); }
a: (x) { fnc1(); }
a: (x) fnc1();

b: (x) -> _ { return fnc2(); }

b: (x) -> _ = fnc2();
b: (x) = fnc2();
```

Also for consistency, other declarations have to be changed:

```cpp
f: (xy) { }
t: type { }
n: namesapce { }

g: (xy) = value;
v: TYPE = value;
```

It's consistent with control flows, `inspect` expressions, and etc.

## Conclusion

To fix this issue, solution 2 is superior to solution 1. Thanks.

**Comment by msadeqhe** [comment]
Now considering `type` and `namespace` declarations, `{}` defines a new one, `=` sets its definition from a previousely defined one:

```cpp
// {} defines a new type:
cls1: type {
    // declarations...
}

// cls2 definition is set from cls1 definition:
cls2: type = cls1;
```

It's similar to `using` statement (type and namespace aliases) in Cpp1.

Considering lambda expressions (unnamed functions), it will be terser without `=` in full lambda syntax:

```cpp
: (x) -> void { print(x); }
: (x) { print(x); }
```

While it's somehow similar to lambdas in Cpp1, but currently we have to write them in Cpp2 (today) with `=`:

```cpp
: (x) -> void = { print(x); }
: (x) = { print(x); }
```

Discussion #793 is related to this topic.
